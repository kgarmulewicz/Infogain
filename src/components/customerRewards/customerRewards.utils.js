import { monthNames } from "../utils";

export const calculateSingleTransactionRewardsAmount = (transaction) => {
  let rewardPoints = 0;
  if (transaction > 50 && transaction <= 100) {
    rewardPoints = rewardPoints + (transaction - 50);
  } else if (transaction > 100) {
    rewardPoints = rewardPoints + (2 * (transaction - 100) + 50);
  }
  return rewardPoints;
};

export const remapData = (customer, yearsAndMonths) => {
  const tempTransactions = [];
  yearsAndMonths.forEach((date) => {
    const tempAmounts = [];
    customer.transactions.forEach((transaction) => {
      if (
        transaction.date.getMonth() === date.month - 1 &&
        transaction.date.getFullYear() === date.year
      ) {
        tempAmounts.push(transaction.amount);
      }
    });
    if (date?.month && tempAmounts.length) {
      tempTransactions.push({
        name: monthNames[date.month - 1],
        transactions: tempAmounts,
      });
    }
  });
  return tempTransactions;
};
