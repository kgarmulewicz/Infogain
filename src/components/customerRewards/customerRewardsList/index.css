.transactions-list {
    width: 100%;
  }