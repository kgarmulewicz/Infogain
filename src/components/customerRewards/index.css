.transactions-container {
  width: 100%;
}

.single-customer {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  width: 100%;
}

.customer-container, .customer-container-active {
    background-color: #eee;
    color: #444;
    cursor: pointer;
    padding: 20px 60px 20px 20px;
    text-align: left;
    border: none;
    outline: none;
    display: flex;
    justify-content: space-between;
  }
  
  .customer-container:after, .customer-container-active:after {
    content: '\02795'; /* Unicode character for "plus" sign (+) */
    font-size: 13px;
    color: #777;
    float: right;
    right: 5px;
    position: absolute;
  }
  
  .customer-container-active:after {
    content: "\2796"; /* Unicode character for "minus" sign (-) */
  }
  
  .customer-container-active, .customer-container:hover {
    background-color: #ccc;
  }
  
  .panel {
    padding: 20px;
    background-color: white;
    color: #a9a9a9;
    overflow: hidden;
    text-align: left;
  }
  
  .months-summary {
    margin: 20px 0;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    width: 100%;
  }
  
  .month-container {
    display: flex;
    align-items: center;
  }
  
  .month-name {
    margin-right: 5px;
  }