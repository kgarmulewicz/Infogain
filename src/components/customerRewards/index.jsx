import React, { useEffect, useMemo, useState } from "react";
import "./index.css";
import CustomerRewardsList from "./customerRewardsList";
import {
  calculateSingleTransactionRewardsAmount,
  remapData,
} from "./customerRewards.utils";

const CustomerRewards = ({ customer, yearsAndMonths }) => {
  const [rewardsPerMonths, setRewardsPerMonths] = useState([]);
  const [isDetailsVisible, setIsDetailsVisible] = useState(false);

  useEffect(() => {
    setRewardsPerMonths(remapData(customer, yearsAndMonths));
  }, [customer, yearsAndMonths]);

  const calculateRewardAmounts = useMemo(() =>
    rewardsPerMonths?.map(({ name, transactions }) => ({
      name,
      monthRewards: transactions.reduce((acc, transaction) =>
        acc + calculateSingleTransactionRewardsAmount(transaction), 0)
    })),
    [rewardsPerMonths],
  );

  const calculateTotalRewardsAmount = useMemo(
    () =>
      calculateRewardAmounts.reduce(
        (acc, { monthRewards }) => acc + monthRewards,
        0
      ),
    [calculateRewardAmounts]
  );

  const toggleDetailsVisibility = () => setIsDetailsVisible(prevState => !prevState);

  return (
    <div data-testid="customer-rewards" className="single-customer">
      <div className="transactions-container">
        <div
          className={
            isDetailsVisible
              ? "customer-container-active"
              : "customer-container"
          }
          onClick={toggleDetailsVisibility}
        >
          <div className="customer-details">
            {`${customer.firstName} ${customer.lastName}`}
          </div>
          <div className="customer-total-rewards">
            {`Total Rewards: ${calculateTotalRewardsAmount}`}
          </div>
        </div>
        <div
          className="panel"
          style={{ display: isDetailsVisible ? "block" : "none" }}
        >
          Customer transactions summary:
          <div className="months-summary">
            {calculateRewardAmounts.map((month) => (
              <div key={month.name} className="month-container">
                <div className="month-name">{month.name}:</div>
                <div>{month.monthRewards}</div>
              </div>
            ))}
          </div>
          <CustomerRewardsList customer={customer} />
        </div>
      </div>
    </div>
  );
};

export default CustomerRewards;
