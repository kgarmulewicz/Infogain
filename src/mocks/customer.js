export const customer = {
            "id": 1,
            "firstName": "Loralee",
            "lastName": "Timeby",
            "transactions": [
                {
                    "date": "2023-01-01T00:00:00.000Z",
                    "amount": 276
                },
                {
                    "date": "2023-01-04T00:00:00.000Z",
                    "amount": 1
                },
                {
                    "date": "2023-01-06T00:00:00.000Z",
                    "amount": 95
                },
                {
                    "date": "2023-01-09T00:00:00.000Z",
                    "amount": 209
                },
                {
                    "date": "2023-01-13T00:00:00.000Z",
                    "amount": 134
                },
                {
                    "date": "2023-01-15T00:00:00.000Z",
                    "amount": 174
                },
                {
                    "date": "2023-01-20T00:00:00.000Z",
                    "amount": 231
                },
                {
                    "date": "2023-01-21T00:00:00.000Z",
                    "amount": 246
                },
                {
                    "date": "2023-01-24T00:00:00.000Z",
                    "amount": 177
                },
                {
                    "date": "2023-01-31T00:00:00.000Z",
                    "amount": 128
                },
                {
                    "date": "2023-02-05T00:00:00.000Z",
                    "amount": 142
                },
                {
                    "date": "2023-02-11T00:00:00.000Z",
                    "amount": 191
                },
                {
                    "date": "2023-02-18T00:00:00.000Z",
                    "amount": 256
                },
                {
                    "date": "2023-02-24T00:00:00.000Z",
                    "amount": 67
                },
                {
                    "date": "2023-02-25T00:00:00.000Z",
                    "amount": 75
                },
                {
                    "date": "2023-02-28T00:00:00.000Z",
                    "amount": 130
                },
                {
                    "date": "2023-03-06T00:00:00.000Z",
                    "amount": 79
                },
                {
                    "date": "2023-03-11T00:00:00.000Z",
                    "amount": 186
                },
                {
                    "date": "2023-03-14T00:00:00.000Z",
                    "amount": 44
                },
                {
                    "date": "2023-03-16T00:00:00.000Z",
                    "amount": 209
                },
                {
                    "date": "2023-03-23T00:00:00.000Z",
                    "amount": 88
                },
                {
                    "date": "2023-03-26T00:00:00.000Z",
                    "amount": 215
                },
                {
                    "date": "2023-03-28T23:00:00.000Z",
                    "amount": 152
                },
                {
                    "date": "2023-03-30T23:00:00.000Z",
                    "amount": 279
                }
            ]
        };