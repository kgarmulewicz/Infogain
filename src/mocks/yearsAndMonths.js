export const yearsAndMonths = [
    {
        "year": 2023,
        "month": 1
    },
    {
        "year": 2023,
        "month": 2
    },
    {
        "year": 2023,
        "month": 3
    }
]