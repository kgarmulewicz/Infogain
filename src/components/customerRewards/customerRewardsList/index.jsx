import React from "react";
import './index.css';
import { formatDate } from "../../utils";
import { calculateSingleTransactionRewardsAmount } from "../customerRewards.utils";

const CustomerRewardsList = ({ customer }) => {
  return (
    <table className="transactions-list">
      <tr>
        <th>Transaction date</th>
        <th>Spent amount</th>
        <th>Reward points</th>
      </tr>
      {customer.transactions.map((transaction) => (
        <tr>
          <td>{formatDate(transaction.date)}</td>
          <td>${transaction.amount}</td>
          <td>{calculateSingleTransactionRewardsAmount(transaction.amount)}</td>
        </tr>
      ))}
    </table>
  );
};

export default CustomerRewardsList;
