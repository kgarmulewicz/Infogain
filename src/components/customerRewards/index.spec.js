import "@testing-library/jest-dom/extend-expect";
import React from "react";
import { render } from "@testing-library/react";
import { customer } from "../../mocks/customer";
import { yearsAndMonths } from "../../mocks/yearsAndMonths";
import CustomerRewards from "./index";
import {calculateSingleTransactionRewardsAmount, remapData} from './customerRewards.utils';

describe("<CustomerRewards />", () => {
  const props = {
    customer,
    yearsAndMonths,
  };
  test("CustomerRewards match snapshot!", () => {
    const { view } = render(<CustomerRewards {...props} />);
    expect(view).toMatchSnapshot();
  });

  test("CustomerRewards calculation method", () => {
    const value = calculateSingleTransactionRewardsAmount(136);
    expect(value).toBe(122);
  });

  test("CustomerRewards maping method", () => {
    const response = remapData(props.customer, props.yearsAndMonths);
    const expectedResponse = [{"name": "January", "transactions": [276, 1, 95, 209, 134, 174, 231, 246, 177, 128]}, {"name": "February", "transactions": [142, 191, 256, 67, 75, 130]}, {"name": "March", "transactions": [79, 186, 44, 209, 88, 215, 152, 279]}];
    expect(response).toStrictEqual(expectedResponse);
  });
});
