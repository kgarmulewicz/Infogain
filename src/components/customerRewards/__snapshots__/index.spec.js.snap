// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomerRewards /> CustomerRewards match snapshot! 1`] = `undefined`;
